code/modules/tgui/tgui_input/list_input.dm

```diff
@@ -0,0 +1,166 @@
+/**
+ * Creates a TGUI input list window and returns the user's response.
+ *
+ * This proc should be used to create alerts that the caller will wait for a response from.
+ * Arguments:
+ * * user - The user to show the input box to.
+ * * message - The content of the input box, shown in the body of the TGUI window.
+ * * title - The title of the input box, shown on the top of the TGUI window.
+ * * items - The options that can be chosen by the user, each string is assigned a button on the UI.
+ * * default - If an option is already preselected on the UI. Current values, etc.
+ * * timeout - The timeout of the input box, after which the menu will close and qdel itself. Set to zero for no timeout.
+ */
+/proc/tgui_input_list(mob/user, message, title = "Select", list/items, default, timeout = 0, ui_state = GLOB.always_state)
+	if(!user)
+		user = usr
+
+	if(!length(items))
+		return
+
+	if(!istype(user))
+		if(!isclient(user))
+			return
+		var/client/client = user
+		user = client.mob
+
+	if(isnull(user.client))
+		return
+
+	/// Client does NOT have tgui_input on: Returns regular input
+	if(user.client?.prefs?.toggles2 & PREFTOGGLE_2_DISABLE_TGUI_INPUT)
+		return input(user, message, title, default) as null|anything in items
+
+	var/datum/tgui_list_input/input = new(user, message, title, items, default, timeout, ui_state)
+
+	if(input.invalid)
+		qdel(input)
+		return
+
+	input.ui_interact(user)
+	input.wait()
+	if(input)
+		. = input.choice
+		qdel(input)
+
+/**
+ * # tgui_list_input
+ *
+ * Datum used for instantiating and using a TGUI-controlled list input that prompts the user with
+ * a message and shows a list of selectable options
+ */
+/datum/tgui_list_input
+	/// The title of the TGUI window
+	var/title
+	/// The textual body of the TGUI window
+	var/message
+	/// The list of items (responses) provided on the TGUI window
+	var/list/items
+	/// Buttons (strings specifically) mapped to the actual value (e.g. a mob or a verb)
+	var/list/items_map
+	/// The button that the user has pressed, null if no selection has been made
+	var/choice
+	/// The default button to be selected
+	var/default
+	/// The time at which the tgui_list_input was created, for displaying timeout progress.
+	var/start_time
+	/// The lifespan of the tgui_list_input, after which the window will close and delete itself.
+	var/timeout
+	/// Boolean field describing if the tgui_list_input was closed by the user.
+	var/closed
+	/// The TGUI UI state that will be returned in ui_state(). Default: always_state
+	var/datum/ui_state/state
+	/// Whether the tgui list input is invalid or not (i.e. due to all list entries being null)
+	var/invalid = FALSE
+
+/datum/tgui_list_input/New(mob/user, message, title, list/items, default, timeout, ui_state)
+	src.title = title
+	src.message = message
+	src.items = list()
+	src.items_map = list()
```

Comment:
These probably don't need to have their lists created here, these really shouldn't be holding out in ram anyways

Reply:
I think I'll keep it just in case

Reply:
This actually may be slower, I'll run a dis to see